src/utils/injectTreeNode.js

```diff
@@ -1,57 +1,54 @@
-import { injectDefaults } from './mobx-react-utils'
 import StripedLoader from './StripedLoader'
-import { autoKey } from './dsl'
-
-export default (
-  render,
-  {
-    type,
-    reactors,
-    nodeProps = _.keys(reactors),
-    allowEmptyNode = false,
-    style,
-  } = {}
-) =>
-  _.flow(
-    inject(({ theme = {} }) => ({
-      // TODO: Update README when the API for Loading components change
-      Loading: (theme.Loading || StripedLoader)(render, style),
-    })),
-    injectDefaults(({ tree, node, group, path, ...props }) => {
-      node = node || tree.getNode(path)
-
-      // Not Found
-      if (!node && path && !allowEmptyNode)
-        throw Error(`Node not found at ${path}`)
-
-      // Dynamic add
-      if (!node && type) {
-        let key = props.nodeKey || autoKey({ type, ...props })
-        group = group || _.get('tree.path', tree)
-
-        // Lookup if already added
-        if (!node) node = tree.getNode([...group, key])
-
-        // Add node if missing
-        if (!node) {
-          let newNode = {
-            key,
-            type,
-            ..._.pick(['field', ...nodeProps], props),
-          }
-          tree.add(group, newNode)
-          // Can't be newNode because it's wrapped in observable, and add doesn't return the new node
-          node = tree.getNode([...group, key])
-
-          if (!node)
-            throw Error(`Unable to add node ${JSON.stringify(newNode)}`)
-        }
-      } else if (!node && !allowEmptyNode)
-        throw Error(`Node not provided, and couldn't find node at ${path}`)
-
-      return { tree, node, isLoading: node && node.updating }
-    })
-  )(({ Loading, ...props }) => <Loading {...props} />)
+
+export let withNode = (Component, { allowEmptyNode = false } = {}) => 
+  ({ tree, node, path, ...props }) => {
+    node = node || tree.getNode(path)
+  
+    if (!node && path && !allowEmptyNode)
+      throw Error(`Node not provided, and couldn't find node at ${path}`)
+      
+    return <Component {...{tree, node, path, ...props}} />
+  }
+
+export let withAutoAdd = (Component, type) => ({ tree, node, group, path, nodeProps, nodeKey, ...props }) => {
+  group = group || path && _.initial(path) || _.get('tree.path', tree)
+  nodeKey = nodeKey || _.last(path)
+
+  // if a key is given, we should see if the node already exists before trying to autogen
+  if (!node && nodeKey) node = tree.getNode([...group, nodeKey])
+  
+  if (!node) {
+    if (!type)
+      throw Error(`Node not provided and cannot be autogenerated ('type' property is required)`)
+  
+    // now that client exposes `types` on the tree, we can look up the type object instead of having to pass it in
+    nodeProps = nodeProps || _.flow(_.get(['types', type, 'reactors']), _.keys)(tree)
+    let newNode = { type, key: nodeKey, ..._.pick(['field', ...nodeProps], props) }
+    // relies on client update to await dispatch and return new node on `tree.add`
+    node = tree.add(group, newNode)
```

Comment:
this PR is probably going to depend on a contexture-client change no matter what, even if it's not a change to `add` like it's assuming here